mtu_discoverer.go

```diff
@@ -25,16 +25,80 @@ const (
 	maxMTUDiff = 20
 	// send a probe packet every mtuProbeDelay RTTs
 	mtuProbeDelay = 5
+	// Once maxLostMTUProbes MTU probe packets larger than a certain size are lost,
+	// MTU discovery won't probe for larger MTUs than this size.
+	// The algorithm used here is resilient to packet loss of (maxLostMTUProbes - 1) packets.
+	maxLostMTUProbes = 3
 )
 
+// The Path MTU is found by sending a larger packet every now and then.
+// If the packet is acknowledged, we conclude that the path supports this larger packet size.
+// If the packet is lost, this can mean one of two things:
+//   1. The path doesn't support this larger packet size, or
+//   2. The packet was lost due to packet loss, independent of its size.
+// The algorithm used here is resilient to packet loss of (maxLostMTUProbes - 1) packets.
+// For simplicty, the following example use maxLostMTUProbes = 2.
+//
+// Initialization:
+//    |------------------------------------------------------------------------------|
+//   min                                                                            max
+//
+// The first MTU probe packet will have size (min+max)/2.
```

Comment:
This follows the `Binary` Candidate Sequence as outlined in the QUIC Path MTU paper. While it performs good, the authors conclude that the `OptBinary` variant is preferable:

> We prefer the optimistic variant of Binary, because it selects
the upper bound of PMTU candidates first and, therewith, has the
potential to immediately either find the best PMTU estimation or
trigger a PTB message. The results of the simulations show that
both algorithms perform well even with external events like data
transmission or packet loss.